tests/test_ops.py

```diff
@@ -186,6 +187,30 @@ async def test_compute_logprob(self, temperature) -> None:
         else:
             assert logp is None
 
+    @pytest.mark.vcr
+    @pytest.mark.asyncio
+    async def test_validation(self) -> None:
+        config = {"model": CommonLLMNames.OPENAI_TEST.value, "num_retries": 1}
+        validator = StatefulValidator()
+        llm_op = LLMCallOp(response_validator=validator)
+        await llm_op(config, msgs=[Message(content="Hello")])
+        assert validator.counter == 2  # first attempt should have failed
+
+        config = {"model": CommonLLMNames.OPENAI_TEST.value, "num_retries": 0}
+        llm_op = LLMCallOp(response_validator=StatefulValidator())
```

Comment:
A simpler way of this is `LLMCallOp(response_validator=MagicMock(side_effect=[ValueError("Invalid response"), None]))`

Basically you can keep things more local to the unit test